### ISSUE TITLE
Decide if we keep SyncOrPromise in the public network API

### ISSUE THREAD

**Issue by kassens** [facebook/relay#1858]
With 18f9d6103f615b622892b4217942db55cb24bdc6 we introduced a `SyncOrPromise` as a type that reflects a value that could either be a promise (guaranteed to be async) or a sync value or error. We need something like this so if a request is fulfilled by a cache or a preloaded network request for example from server side rendering, that we can immediately render the loaded state and don't waste cycles on rendering a loading state first.

@josephsavona brought up a good point that we might want to avoid this as part of the public API and maybe just allow a callback that can be called synchronously or async.

cc @taion, @JenniferWang

**Comment by taion** [comment]
This seems to be a breaking change – a `fetch` function that currently just returns a normal promise won't work any more.

Otherwise I like and have used the pattern in graphql-js of taking both promise and non-promise values, and handling them both appropriately: https://github.com/graphql/graphql-js/blob/v0.10.1/src/execution/execute.js#L359-L366

**Comment by kassens** [comment]
I like this because it keeps the API compatible even if slightly less explicit about what's being returned.

**Comment by josephsavona** [comment]
Thanks for opening this. My main concerns with SyncOrPromise are:
- it's non-standard and hard to use (the name suggests you can return a promise but you have to wrap it)
- it's not composable, meaning that everything that consumes such a value becomes more complicated. For example you can "map" a Promise with .then(), but mapping a SyncOrPromise involves handling two different cases that are very different. 

What we really need is a monadic type that represents a value that can be resolved synchronously or asynchronously. The only way to achieve that is with functions, aka using a class (object with methods) as opposed to a plain data-only object. It also must auooort cancellation. That leaves:
- Observable (possibly with onNext wrapped to prevent returning more than one value
- Custom callbacks with an optional Promise return value. This might look like `fetch(op, vars, cacheConfig, observer): (Disposable | Promise)`. The implementation must either return a Promise *or* return a disposable  and call the observer's resolve/reject functions. 
- A custom type, such as the `Single` type I created for this exact use-case in [rsocket-js](https://github.com/rsocket/rsocket-js/blob/master/docs/03-flowable-api.md). 

If we were starting from scratch I would suggest the latter, given the current state the middle option (pass an observer and accept an optional Promise response) seems good.

**Comment by josephsavona** [comment]
I forgot to mention: it's useful to separate out the type of the *internal* network layer fro what is accepted by Network.create(). Internally it makes sense to use something like the `Single` type I mentioned Veda's of composition and consistency. Network.create() can use the intermediate approach of converting Promises to the internal type.

**Comment by taion** [comment]
This is sort of a fuzzy semver boundary thing – while the old typing was `Promise<RelayResponsePayload>`, most likely things worked with a return type of `Promise<RelayResponsePayload> | RelayResponsePayload`, so it'd be a clean API in many cases to continue to support both.

Also, the full intersection of features probably isn't required – for example, cancellation doesn't make sense for an eager response.

One option might be:

```js
(Promise<Result> & Disposable) | Promise<Result> | Result;
```

So in other words, if it doesn't have `.then()`, it's a sync response; otherwise check whether it has `.dispose()`. The former part is exactly what graphql-js does; the latter part seems a straightforward extension.

**Comment by taion** [comment]
So this is a "no", right?

**Comment by wincent** [comment]
Thanks for the input here. We're currently going through old issues that appear to have gone stale (ie. not updated in about the last 3 months) because the volume of material in the issue tracker is becoming hard to manage. If this is still important to you please comment and we'll re-open this.

Thanks once again!
